### ISSUE TITLE
Provide a better API for decoding a token without signature validation

### ISSUE THREAD

**Issue by tyilo** [Keats/jsonwebtoken#401]
Currently you have to use:
```rust
// Algorithm can be arbitrarily chosen
let mut validation = jsonwebtoken::Validation::new(jsonwebtoken::Algorithm::RS256);
validation.insecure_disable_signature_validation();

// Key can be arbitrarily chosen
let key = DecodingKey::from_secret(&[]);
let payload = jsonwebtoken::decode::<Claims>(token, &key, &validation).unwrap();
```

I think the following API would be better:
```rust
let mut validation = jsonwebtoken::Validation::insecure_without_signature_validation();
let payload = jsonwebtoken::insecure_decode_without_signature_validation::<Claims>(token, &validation).unwrap();
```

You avoid having to choose a random algorithm and decoding key that isn't ever used.

**Comment by Keats** [comment]
Honestly, decoding a token without validating the signature is something that you shouldn't do most of the time so I do not particularly care about making it user friendly

**Comment by tyilo** [comment]
> Honestly, decoding a token without validating the signature is something that you shouldn't do most of the time so I do not particularly care about making it user friendly

It is useful as a client using the token to be able to see what claims are inside the token. `exp` can be really useful.

**Comment by Keats** [comment]
Well you can't really trust any of the things you see in the claims unless you validate the signature

**Comment by pooriatgh** [comment]
Today, I encountered the same issue. It would be nice if we could make this easier. Here is the scenario:

1. I need to decode the token, look into the claims without validation, and extract the issuer's name.
2. Use the issuer's name to find the correct public key file from a list of public keys I have.
3. Use the key to perform the verification/validation.

I tried the solution provided by @tyilo , but I failed to decode without the audience validation step. I used this option to disable it:


```rust
let mut validation = jsonwebtoken::Validation::new(jsonwebtoken::Algorithm::RS256);
validation.insecure_disable_signature_validation();
validation.validate_aud = false;
```

For example in typescript this package provide more or less what I need: [node-jsonwebtoken](https://github.com/auth0/node-jsonwebtoken)

**Comment by ppamorim** [comment]
> > Well you can't really trust any of the things you see in the claims unless you validate the signature
> 
> Sure, but I'm the client. I don't have access to the server's secret key.

You are totally right, it makes no sense to not be able to decode JWT in the client side, this function is pretty simple and supported by most of the other JWT libraries I had to face. When this library is used on a server side it totally makes sense to force the validation as it is basically how JWT works.

There is really no real benefit in not implementing this feature and the author is asking for the impossible, we cannot use this library without those workarounds.

**Comment by MichaelRollins** [comment]
Any update on this? I have the same requirement as @pooriatgh.

**Comment by Nereuxofficial** [comment]
Would it be acceptable to hide that API behind a .dangerous function call similar to how [rustls](https://docs.rs/rustls/latest/rustls/struct.ConfigBuilder.html#method.dangerous) implements such things. If that would be an acceptable compromise i could create a MR for it.

**Comment by Keats** [comment]
I'd take a PR for something like `jsonwebtoken::insecure_decode_without_signature_validation`
